core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -547,12 +547,136 @@ public CompletableFuture<Optional<Throwable>> acknowledge(
      * @return A future which is completed when the records are released.
      */
     public CompletableFuture<Optional<Throwable>> releaseAcquiredRecords(String memberId) {
-        log.trace("Release acquired records request for share partition: {}-{}-{}", groupId, memberId, topicIdPartition);
+        log.trace("Release acquired records request for share partition: {}-{} memberId: {}", groupId, topicIdPartition, memberId);
 
-        CompletableFuture<Optional<Throwable>> future = new CompletableFuture<>();
-        future.completeExceptionally(new UnsupportedOperationException("Not implemented"));
+        Throwable throwable = null;
+        lock.writeLock().lock();
+        List<InFlightState> updatedStates = new ArrayList<>();
+        List<PersisterStateBatch> stateBatches = new ArrayList<>();
+
+        try {
+            RecordState recordState = RecordState.AVAILABLE;
+            // Iterate over multiple fetched batches. The state can vary per offset entry
+            for (Map.Entry<Long, InFlightBatch> entry : cachedState.entrySet()) {
+                InFlightBatch inFlightBatch = entry.getValue();
+
+                if (inFlightBatch.offsetState() == null
+                        && inFlightBatch.batchState() == RecordState.ACQUIRED
+                        && inFlightBatch.batchMemberId().equals(memberId)
+                        && checkForStartOffsetWithinBatch(inFlightBatch.firstOffset(), inFlightBatch.lastOffset())) {
+                    // For the case when batch.firstOffset < start offset <= batch.lastOffset, we will be having some
+                    // acquired records that need to move to archived state despite their delivery count.
+                    inFlightBatch.maybeInitializeOffsetStateUpdate();
+                }
+
+                if (inFlightBatch.offsetState() != null) {
+                    throwable = releaseAcquiredRecordsForPerOffsetBatch(memberId, inFlightBatch, recordState, updatedStates, stateBatches);
+                    if (throwable != null)
+                        break;
+                    continue;
+                }
+                throwable = releaseAcquiredRecordsForCompleteBatch(memberId, inFlightBatch, recordState, updatedStates, stateBatches);
+                if (throwable != null) {
+                    break;
+                }
+            }
+
+            // If the release acquired records is successful then persist state, complete the state transition
+            // and update the cached state for start offset. Else rollback the state transition.
+            rollbackOrProcessStateUpdates(throwable, updatedStates, stateBatches);
+        } finally {
+            lock.writeLock().unlock();
+        }
+        return CompletableFuture.completedFuture(Optional.ofNullable(throwable));
```

Comment:
The `CompletableFuture<Optional<Throwable>>` is a bit unusual. Generally, a future with a Throwable would complete exceptionally, rather than completing with a result which is an optional Throwable. Then, instantiating an exception is an expensive operation which captures a stack trace and so on. So, I wonder whether this is really the best pattern for passing back the results from these methods.

Reply:
Since, we need to use the exception to return to the API that functions `acknowledge` and `releaseAcquiredRecords` return, I think we can go ahead with option 2. The tradeoffs remain the same with both option 2 and option 3 as mentioned by @apoorvmittal10 .